dbms/src/Storages/MergeTree/MergeTreeRangeReader.cpp

```diff
@@ -627,40 +649,74 @@ Block MergeTreeRangeReader::continueReadingChain(ReadResult & result)
         if (next_range_to_start < started_ranges.size()
             && i == started_ranges[next_range_to_start].num_granules_read_before_start)
         {
-            added_rows += stream.finalize(block);
+            num_rows += stream.finalize(columns);
             auto & range = started_ranges[next_range_to_start].range;
             ++next_range_to_start;
             stream = Stream(range.begin, range.end, merge_tree_reader);
         }
 
         bool last = i + 1 == size;
-        added_rows += stream.read(block, rows_per_granule[i], !last);
+        num_rows += stream.read(columns, rows_per_granule[i], !last);
     }
 
     stream.skip(result.numRowsToSkipInLastGranule());
-    added_rows += stream.finalize(block);
+    num_rows += stream.finalize(columns);
 
     /// added_rows may be zero if all columns were read in prewhere and it's ok.
-    if (added_rows && added_rows != result.totalRowsPerGranule())
-        throw Exception("RangeReader read " + toString(added_rows) + " rows, but "
+    if (num_rows && num_rows != result.totalRowsPerGranule())
+        throw Exception("RangeReader read " + toString(num_rows) + " rows, but "
                         + toString(result.totalRowsPerGranule()) + " expected.", ErrorCodes::LOGICAL_ERROR);
 
-    return block;
+    return columns;
 }
 
 void MergeTreeRangeReader::executePrewhereActionsAndFilterColumns(ReadResult & result)
 {
     if (!prewhere_actions)
         return;
 
-    if (alias_actions)
-        alias_actions->execute(result.block);
+    auto & header = merge_tree_reader->getColumns();
+    size_t num_columns = header.size();
+
+    if (result.columns.size() != num_columns)
+        throw Exception("Invalid number of columns passed to MergeTreeRangeReader. "
+                        "Expected " + toString(num_columns) + ", "
+                        "got " + toString(result.columns.size()), ErrorCodes::LOGICAL_ERROR);
 
-    prewhere_actions->execute(result.block);
-    auto & prewhere_column = result.block.getByName(*prewhere_column_name);
-    size_t prev_rows = result.block.rows();
-    ColumnPtr filter = prewhere_column.column;
-    prewhere_column.column = nullptr;
+    ColumnPtr filter;
+    size_t prewhere_column_pos;
+
+    {
+        /// Restore block from columns list.
+        Block block;
+        size_t pos = 0;
+
+        if (prev_reader)
+        {
+            for (auto & col : prev_reader->getSampleBlock())
+            {
+                block.insert({result.columns[pos], col.type, col.name});
+                ++pos;
+            }
+        }
+
+        for (auto name_and_type = header.begin(); pos < num_columns; ++pos, ++name_and_type)
+            block.insert({result.columns[pos], name_and_type->type, name_and_type->name});
+
+        if (alias_actions)
+            alias_actions->execute(block);
```

Comment:
In the future we will execute all actions on `header` and `columns`?

Reply:
Yes, I suppose so.